### ISSUE TITLE
Substream Source  cannot be materialized more than once in GrpchMarshalling

### ISSUE THREAD

**Issue by marekzebrowski** [akka/akka-grpc#886]
I have a very simple request-response grpc service that started to fail after upgrade to akka-grpc 0.8.1
actual service implementation is rather boring, but something is wrong in marshalling part.
generated code that matches following stacktrace looks like:
```scala
  GrpcMarshalling.unmarshal(request.entity.dataBytes)(AddEmojiRequestSerializer, mat, reader)
                  .flatMap(implementation.addEmoji(_))
                  .map(e => GrpcMarshalling.marshal(e, eHandler)(AddEmojiResponseSerializer, writer, system))
```
so nothing fancy here, yet it fails

### Versions used 
Akka version: 2.5.30
akka-http: 10.1.11
akka-grpc: 0.8.1
Scala 2.13.1
### Expected Behavior
Response is serialized properly - that worked fine up to akka-grpc 0.7.3

### Relevant logs
```
14:34:12.223 [emojis-akka.actor.default-dispatcher-5] ERROR akka.actor.ActorSystemImpl - Unhandled error: [Substream Source(substream-out-3) cannot be materialized more than once].
java.lang.IllegalStateException: Substream Source(substream-out-3) cannot be materialized more than once
	at akka.stream.impl.fusing.SubSource$$anon$13.createMaterializedTwiceException(StreamOfStreams.scala:830)
	at akka.stream.impl.fusing.SubSource$$anon$13.<init>(StreamOfStreams.scala:801)
	at akka.stream.impl.fusing.SubSource.createLogic(StreamOfStreams.scala:797)
	at akka.stream.stage.GraphStage.createLogicAndMaterializedValue(GraphStage.scala:107)
	at akka.stream.stage.GraphStageWithMaterializedValue.createLogicAndMaterializedValue(GraphStage.scala:51)
	at akka.stream.impl.GraphStageIsland.materializeAtomic(PhasedFusingActorMaterializer.scala:675)
	at akka.stream.impl.PhasedFusingActorMaterializer.materialize(PhasedFusingActorMaterializer.scala:492)
	at akka.stream.impl.PhasedFusingActorMaterializer.materialize(PhasedFusingActorMaterializer.scala:451)
	at akka.stream.impl.PhasedFusingActorMaterializer.materialize(PhasedFusingActorMaterializer.scala:444)
	at akka.stream.scaladsl.RunnableGraph.run(Flow.scala:629)
	at akka.stream.scaladsl.Source.runWith(Source.scala:106)
	at akka.grpc.scaladsl.GrpcMarshalling$.unmarshal(GrpcMarshalling.scala:52)
	at me.sgrouples.emojis.grpc.EmojiServiceHandler$.$anonfun$partial$1(EmojiServiceHandler.scala:86)
	at akka.grpc.scaladsl.GrpcMarshalling$.$anonfun$negotiated$2(GrpcMarshalling.scala:44)
	at scala.util.Success.map(Try.scala:262)
	at akka.grpc.scaladsl.GrpcMarshalling$.$anonfun$negotiated$1(GrpcMarshalling.scala:44)
	at scala.Option.map(Option.scala:242)
	at akka.grpc.scaladsl.GrpcMarshalling$.negotiated(GrpcMarshalling.scala:42)
	at me.sgrouples.emojis.grpc.EmojiServiceHandler$.handle$1(EmojiServiceHandler.scala:81)
	at me.sgrouples.emojis.grpc.EmojiServiceHandler$.$anonfun$partial$21(EmojiServiceHandler.scala:145)
	at scala.PartialFunction$Unlifted.applyOrElse(PartialFunction.scala:319)
	at scala.PartialFunction$OrElse.apply(PartialFunction.scala:238)
	at akka.grpc.scaladsl.ServiceHandler$$anonfun$handler$1.applyOrElse(ServiceHandler.scala:49)
	at akka.grpc.scaladsl.ServiceHandler$$anonfun$handler$1.applyOrElse(ServiceHandler.scala:48)
	at scala.PartialFunction$OrElse.apply(PartialFunction.scala:238)
	at akka.http.impl.engine.http2.Http2Blueprint$.$anonfun$handleWithStreamIdHeader$1(Http2Blueprint.scala:123)
	at akka.stream.impl.fusing.MapAsyncUnordered$$anon$31.onPush(Ops.scala:1376)
	at akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:523)
	at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:409)
	at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:606)
	at akka.stream.impl.fusing.ActorGraphInterpreter$SimpleBoundaryEvent.execute(ActorGraphInterpreter.scala:47)
	at akka.stream.impl.fusing.ActorGraphInterpreter$SimpleBoundaryEvent.execute$(ActorGraphInterpreter.scala:43)
	at akka.stream.impl.fusing.ActorGraphInterpreter$BatchingActorInputBoundary$OnNext.execute(ActorGraphInterpreter.scala:85)
	at akka.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:581)
	at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:749)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:764)
	at akka.actor.Actor.aroundReceive(Actor.scala:539)
	at akka.actor.Actor.aroundReceive$(Actor.scala:537)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:671)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:612)
	at akka.actor.ActorCell.invoke(ActorCell.scala:581)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:268)
	at akka.dispatch.Mailbox.run(Mailbox.scala:229)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:241)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by marekzebrowski** [comment]
I tried the same thing with akka 2.6.4 and similar results:
```
java.lang.IllegalStateException: Substream Source(substream-out-3) cannot be materialized more than once
	at akka.stream.impl.fusing.SubSource$$anon$13.createMaterializedTwiceException(StreamOfStreams.scala:826)
	at akka.stream.impl.fusing.SubSource$$anon$13.<init>(StreamOfStreams.scala:797)
	at akka.stream.impl.fusing.SubSource.createLogic(StreamOfStreams.scala:793)
	at akka.stream.stage.GraphStage.createLogicAndMaterializedValue(GraphStage.scala:105)
	at akka.stream.stage.GraphStageWithMaterializedValue.createLogicAndMaterializedValue(GraphStage.scala:49)
	at akka.stream.impl.GraphStageIsland.materializeAtomic(PhasedFusingActorMaterializer.scala:700)
	at akka.stream.impl.PhasedFusingActorMaterializer.materialize(PhasedFusingActorMaterializer.scala:499)
	at akka.stream.impl.PhasedFusingActorMaterializer.materialize(PhasedFusingActorMaterializer.scala:449)
	at akka.stream.impl.PhasedFusingActorMaterializer.materialize(PhasedFusingActorMaterializer.scala:441)
	at akka.stream.scaladsl.RunnableGraph.run(Flow.scala:703)
	at akka.stream.scaladsl.Source.runWith(Source.scala:109)
	at akka.grpc.scaladsl.GrpcMarshalling$.unmarshal(GrpcMarshalling.scala:52)
	at me.sgrouples.emojis.grpc.EmojiServiceHandler$.$anonfun$partial$1(EmojiServiceHandler.scala:86)
```

**Comment by raboof** [comment]
Thank you for your report! That definitely looks suspicious. I'll look into it.

**Comment by marekzebrowski** [comment]
I suspect changes introduced by [Grpc-web support](https://github.com/akka/akka-grpc/commit/ba0793c4ff02bd73b81e24530e5a9aa373a05354#diff-f3c23e9cbdb39a39817c6102ac2b4996L19) but I'm not 100% sure yet

**Comment by raboof** [comment]
> I suspect changes introduced by Grpc-web support but I'm not 100% sure yet

That sounds likely as that's the largest materialization-related change, but so far I haven't been able to reproduce the problem. Would it be possible for you to boil your application down to something that still shows the problem and you can share?

**Comment by marekzebrowski** [comment]
OK, I'll try to minimize the problem and come back with a reproducer

**Comment by marekzebrowski** [comment]
I have a reproducer https://github.com/marekzebrowski/grpc-bug-reproducer
it seems that having multiple uses of `Http2.bindAndHandle` triggers the problem - see https://github.com/marekzebrowski/grpc-bug-reproducer/blob/master/src/main/scala/Server.scala#L33

If I have only one `bindAndHandle` it works fine (or with grpc 0.7.x)

**Comment by marekzebrowski** [comment]
After further check - that second bindAndHandle can be used for something else - like status route and it still triggers the error

**Comment by raboof** [comment]
That's fascinating! I can't explain that yet, hope to dig in later this week. Thanks a lot for the reproducer!

**Comment by raboof** [comment]
(I identified the problem, will work on a fix)

**Comment by raboof** [comment]
(should be fixed in 0.8.4)
